### ISSUE TITLE
One more call for stemming. 

### ISSUE THREAD

**Issue by buriy** [explosion/spaCy#4815]
First of all, there's a problem (in Russian):
After training, you should have 
```
POS('кодила') -> VERB (translation: [she] coded)
POS('крокодила') -> NOUN (translation: [of] crocodile)
```
But looking at prefix (first char) and suffix (last 3 chars), if you haven't met these exact words in the training corpus... they are the same for the model! Both are VERBs or both are NOUNs.
Looking at examples in which crocodiles are VERBS makes me laugh... Have you crocodiled somebody yesterday? :)
Why you might not meet the words in training? Because we have:
```
len(russian_dictionary) ~ 5,000,000
len(russian_lemmas) ~ 500,000
len(russian_stems) ~ 550,000
24 words/lemma for ADJs,
10-12 words/lemma for NOUNs,
15-20 words/lemma for VERBs
```
But stemming would work well.
Now, remembering the Zipf law definition, given that we have HashEmbeds of 
 ( orth | lower | norm | prefix | suffix | shape ) + G(vectors) + H(tok2vec),
you still have no chance to meet every word even once(!) in training, and to recognize them correctly in production then. Pretraining vectors and then sharing same vectors for all words for the same stem... well, you'll still have at least 1-2% of OOV words even with 100 GBs of text. (And two tables of 5M links +500K vectors to load...)
However, you could often get POS from the word stem, and that should improve POS accuracy from 93% as we now have in https://github.com/buriy/spacy-ru , to more typical 97-98%. Syntax and NER quality will be improved too.

For one experiment, I've got 65% quality with TextCat for news category guessing, but quality increased to 72% when I just replaced words with their first sense lemmas from pymorphy2. We typically get the same result with stems, which are just words without their suffixes for Russian Snowball algorithm.
With hash-storage, no need to worry about that stem won't perfectly match the lemma, we only need to have much less stems than words so that combination of [prefix + stem + word + lower + shape] would be more powerful for rare words than the same without stems.

So at the moment I'm considering replacing .lower or .norm with stems using a custom tokenizer or some other dirty hack... but I'd like to include Russian fast and near-SOTA models to Spacy, and with that goal in mind, it's not a good option, right?
Another workaround is to replace randomly one name with another name, one loc with another loc -- that can be done for NERs while training, but it's complicated and won't work for less frequent regular words.

So, I think we need this for morphology-rich languages that have a lot of unpredictable word POS from only prefix and suffix: Russian, Ukrainian, Bulgarian, a lot of other slavic languages, then I guess German, Turkish.
Even French and English should get OOV improvements, but I think not more than 0.1%, based on much lower probabilities of OOV when one has fewer words in the vocabulary.
You might also see improvements with Prodigy active-learning even without vectors / tok2vec, it would be interesting to try that.

Related issues: #327 #1825 
Alternative solution is proper support for fastText vectors: #2154 #3978

**Comment by buriy** [comment]
Meanwhile, this monkey-patching works (65% -> 70% on my simple test example), and I can put it into Lemmatizer for Russian language.
@honnibal @ines 
Do you think it is ok to do it in this way? Will you accept this way of solving the problem?

```
# pip install pystemmer
import Stemmer

def change_tokenizer(nlp):
    tok = nlp.tokenizer
    stemmer = Stemmer.Stemmer('ru')
    def stemming_tokenizer(tokens):
        tokens = tok(tokens)
        for t in tokens:
            t.norm_ = stemmer.stemWord(t.norm_)
        return tokens
    nlp.tokenizer = stemming_tokenizer
```

**Comment by buriy** [comment]
Also specifically mentioning #2735 where @honnibal tells that rewriting token.norm_ is fine for this purpose.

**Comment by BramVanroy** [comment]
The downside that I see is that you are then adding another dependency whereas efforts have been made to minimize third-party dependencies.

**Comment by buriy** [comment]
@BramVanroy 
 We can use pymorphy2 instead (which is already a prerequisite) or just removing 3 last characters -- if only this worked!
Currently, it doesn't work well for POS+DEP training for some reason, but works for classification. 
So now I'm considering a better approach of using Morphologizer + char_embed (I guess these are prepared for spacy 2.3 -- no docs, a switch turned off by default, but the code is ready in Spacy ).

**Comment by svlandeg** [comment]
Hi @buriy, apologies for the late follow-up, and thanks for the detailed write-up! 

I definitely see your point, but I think that this kind of feature would be best implemented as an external plugin. From spaCy v.3 onwards you'll also have much more flexibility for implementing your own models (with custom features) through a training config file.

If you do end up writing this plugin, please feel free to add it to the [spacy universe](https://spacy.io/universe)! I'll close this issue in the meantime.

**Comment by buriy** [comment]
Hi @svlandeg , and is fasttext vectors support in roadmap? Is compressed fasttext vectors support in roadmap?
My purpose here is contributing high-quality Russian models to the Spacy model pool, not just using those for my own, so I'm looking for a solution that won't involve mangling spacy code. Or I can contribute code to spacy.lang.ru if that would be easier.

**Comment by svlandeg** [comment]
Yea, you can use `spacy init-model` to create a basic model from external vectors such as from fasttext - see https://spacy.io/usage/vectors-similarity#converting

**Comment by buriy** [comment]
@svlandeg if you read the text on top ^^^ you would understand why the current solution won't work well:
len(russian_dictionary) ~ 5,000,000
len(russian_lemmas) ~ 500,000
len(russian_stems) ~ 550,000
Even with 10x pruning/quantization, it's still 500k vectors of length 300, which is 500k*300*4 = 600 MB !
To store all 5M vectors, you need 6 GB.
Good vectors store both morpho attributes (partially covered by prefix and suffix classifiers) and syntactic-semantic attributes.
Naive quantization would mess it up when pruning.

Right now I managed to store top-250k vectors of length 300 created with manual vectors quantization algorithm based on lemmas, but it's still 300 MB for the smallest Russian model, and 600 MB for a larger one. And we store 5M hashes in the key2row. 
After conversion, fasttext n-grams are removed, and spacy prefix and suffix tables are learned instead. This works, but not as good as it could. We still need to provide more morphology clues to the model, but there's no way to do that.

Also, there are advanced quantization techniques, implemented in https://pypi.org/project/compress-fasttext/ which give additional 5x-10x compression for the vectors (totalling to 50x-100x), allowing to safely store vectors for much more rare words and still get into 100 MB on disk and 600 MB in memory.

**Comment by adrianeboyd** [comment]
I think the confusion/miscommunication here is related to fasttext vectors exported as word2vec-style vectors for full tokens vs. fasttext models with character n-gram features.

`CharEmbed` will be a good option in spacy v3. Be aware that you're probably going to run into bugs in models with `CharEmbed` on GPU in v2.3, so we'd recommend waiting for v3.

Longer term (this would come after the v3 release), we'd like to have the option to replace `HashEmbed` with vectors based on fasttext character n-gram vectors, so you can have vectors for every token from a relatively small vector table. The idea is to use only the char n-gram table while training in fasttext, but with Bloom embeddings so you can scale down the size of the table (the default of 2M is huge). I've implemented the fasttext side of things (https://github.com/adrianeboyd/fastText/tree/feature/bloom), but it still needs to be integrated with thinc and spacy.

**Comment by buriy** [comment]
Hi Adriane, but... tokenization is different in fasttext... Maybe, it is better to add better pre-training right into spaCy? 
Copying/reusing a part of fasttext code that does the training should be easy, but Gensim proved once that you can do the same in python and do it rather fast.
I'm not sure that we only should use n-grams for fasttext training, because pre-trained word vectors are also useful (for top words).
I think It's best to pre-train exactly tok2vec for all purposes, only you need to make it work fast and make it customizable for other languages (which is almost done I guess).
I many times suggested the best working version for Russian and many other languages, which wouldn't take much embedding table space and will work fast: shape vector + lemma vector + morphology vector, where lemma is taken from a big table (word -> lemma). The vectors can be learned, of course, so one only needs two dictionaries: word -> lemma id and word -> morphology id. Note that there could be several candidates both for lemma and morphology. They can be averaged, I think. For German... instead of candidates, you could use a lemma and morphology for each word stem.
I would need to try all that someday with multiple languages.

Also, BPE pre-training should work even better now in 2020 (especially with BPE-dropout): this should work much faster than adding vectors for all n-grams while having almost the same quality.
It's like:
for <10k samples: word embeddings
for 10k-1M samples: word embeddings + n-grams embeddings
for 1M+ samples: BPE embeddings.
(..or you could train BPE embeddings with pre-training.)
